interfaces/cython/cantera/composite.py

```diff
@@ -655,6 +656,33 @@ def __call__(self, *species):
         return SolutionArray(self._phase[species], states=self._states,
                              extra=self._extra)
 
+    def __len__(self) -> int:
+        return self._shape[0]
+
+    @property
+    def ndim(self) -> int:
+        """The number of dimensions in the SolutionArray.
+
+        :return: `int`
+        """
+        return len(self.shape)
+
+    @property
+    def shape(self) -> tuple[int, ...]:
```

Comment:
Should likely be `Tuple` (imported from `typing`)

Reply:
I _think_ I did the `from __future__ import annotations` incantation at the top... if not, that's my preference, since we can eventually drop the import.

Reply:
I’d be ok with that, as long as unit tests pass …